Detailed Card

mashrukh-islam opened the pull request:
### Ticket Link
---------------------------------------------------


### Related Links
---------------------------------------------------


### Description
---------------------------------------------------


### Steps to Reproduce / Test
---------------------------------------------------


### GIF's
---------------------------------------------------
![DetailCard](https://user-images.githubusercontent.com/87057006/125601076-ae8c46a1-6e40-455d-bbee-1f059f2fde42.gif)

mashrukh-islam committed:
detailed track card

An ITunesProvider container is created using floki. This houses a TrackContainer,
to display a list of tracks to the user based on the search query and a TrackDetailsContainer,
that displays details of specific track when the user clicks 'See More' button on the track.
The TrackDetailsContainer first tries to fetch the track information from the store, if the data
doesn't exist in the store, the container makes an API call and fetches the data. Three Card Actions are
provided, which link to the arsists's page, albums' page and tracks' page respectively.

mashrukh-islam committed:
Removed stray strings from Card Actions

alichherawalla commented on app/containers/ITunesProvider/TrackDetailsContainer/index.js:
```diff
@@ -0,0 +1,141 @@
+}) {
+  const [loading, setLoading] = useState(true);
+  useInjectSaga({ key: 'iTunesProvider', saga });
+  const params = useParams();
```
destructure here itself
so 
```
const { trackId } = useParams(); 
```

alichherawalla commented on app/containers/ITunesProvider/TrackDetailsContainer/index.js:
```diff
@@ -0,0 +1,141 @@
+      >
+        <Row>
+          <Col size={1}>
+            <Statistic title="Track Name" value={trackDetails.trackCensoredName} />
```
use react-intl, lets make it a practice. No stray strings

alichherawalla commented on app/containers/ITunesProvider/TrackDetailsContainer/index.js:
```diff
@@ -0,0 +1,141 @@
+  dispatchFetchTrackDetails: PropTypes.func,
+  dispatchClearTrackDetails: PropTypes.func,
+  trackId: PropTypes.string,
+  trackDetails: PropTypes.object,
```
its not always needed but it'll be good to do it this one time. Lets actually chart out the shape of the object using PropTypes.shape

alichherawalla commented on app/containers/ITunesProvider/TracksContainer/index.js:
```diff
@@ -0,0 +1,140 @@
+  tracksError
+}) {
+  useInjectSaga({ key: 'iTunesProvider', saga });
+  const [loading, setLoading] = useState(false);
```
do we need another loading state variable? 

There is a "loader" that we can read from the redux store yes?

alichherawalla replied:
@mashrukh-islam

alichherawalla commented on app/containers/ITunesProvider/index.js:
```diff
@@ -0,0 +1,46 @@
+import { makeSelectITunesProvider } from './selectors';
+import saga from './saga';
+
+export function ITunesProvider() {
```
we can delete this right?

alichherawalla commented on app/containers/ITunesProvider/saga.js:
```diff
@@ -0,0 +1,42 @@
+export function* fetchTrackDetails(action) {
+  const track = yield select(selectTrackById(action.trackId));
+  if (track) {
+    yield put(successGetTrackDetails(track));
```
this doesn't really append to the list, so it doesn't handle the case that we discussed today

mashrukh-islam committed:
Fixed issues and bugs according to review

mashrukh-islam committed:
Tests for tracks reducer

Tests are written for the tracks reducer. Test cases are REQUEST_GET_TRACK_NAMES,
SUCCESS_GET_TRACK_NAMES,CLEAR_TRACK_NAMES. All the test cases are passing.

alichherawalla commented on app/containers/ITunesProvider/TrackDetailsContainer/index.js:
```diff
@@ -0,0 +1,170 @@
+  dispatchFetchTrackDetails: PropTypes.func,
+  dispatchClearTrackDetails: PropTypes.func,
+  trackDetails: PropTypes.shape({
+    artistId: PropTypes.number,
```
some out of these will be required right?

alichherawalla commented on app/containers/ITunesProvider/reducer.js:
```diff
@@ -0,0 +1,58 @@
+        break;
+      case iTunesProviderTypes.CLEAR_TRACK_DETAILS:
+        return initialState;
+      case iTunesProviderTypes.SUCCESS_GET_TRACK_DETAILS:
```
draft.trackDetails = action.data;
draft.tracks = [...state.tracks, action.data];

alichherawalla commented on app/containers/ITunesProvider/saga.js:
```diff
@@ -0,0 +1,43 @@
+    const { data, ok } = response;
+    if (ok) {
+      yield put(successGetTrackDetails(data.results[0]));
+      yield put(successGetTrackNames(data.results[0]));
```
remove this line

alichherawalla commented on app/containers/ITunesProvider/reducer.js:
```diff
@@ -0,0 +1,58 @@
+        draft.tracks = action.data;
+        break;
+      case iTunesProviderTypes.FAILURE_GET_TRACK_NAMES:
+        draft.tracksError = get(action.error, 'message', 'something_went_wrong');
```
get(action.error, 'message', translate('something_went_wrong'));

alichherawalla replied:
move to saga

mashrukh-islam committed:
Completed test cases for iTunesProvider reducer

mashrukh-islam committed:
Completed tests for saga and selectors

Tests are written for saga and selectors

mashrukh-islam committed:
Loader from redux store and miscellaneous changes

alichherawalla commented on app/containers/ITunesProvider/saga.js:
```diff
@@ -0,0 +1,47 @@
+import get from 'lodash/get';
+import { selectTrackById, selectTracksData } from './selectors';
+import { translate } from '@app/components/IntlGlobalProvider/index';
+// Individual exports for testing
```
remove

mashrukh-islam committed:
Removed translate from saga and placed in component

mashrukh-islam committed:
Finished saga and reducer tests

mashrukh-islam committed:
intlProvider for saga tests

mashrukh-islam committed:
Replaced loading variable with store data

Earlier used loading varialbe from store data, is replaced with
loading variable from the store

alichherawalla commented on app/containers/ITunesProvider/tests/saga.test.js:
```diff
@@ -0,0 +1,68 @@
+    );
+  });
+
+  it('should ensure that the action SUCCESS_GET_TRACK_DETAILS is dispatched when the store call succeeds', () => {
```
test both cases here. The one where we are able to fulfil the request from the redux store and the case where we fulfil it from the API

mashrukh-islam replied:
Yea implemented this.

alichherawalla commented on app/containers/ITunesProvider/tests/saga.test.js:
```diff
@@ -0,0 +1,68 @@
+  it('should ensure that the action SUCCESS_GET_TRACK_DETAILS is dispatched when the store call succeeds', () => {
+    getTrackDetailsGenerator = fetchTrackDetails({ trackId });
+    const res = getTrackDetailsGenerator.next().value;
+    expect(JSON.stringify(res)).toEqual(JSON.stringify(select(selectTrackById(trackId))));
```
you don't need to stringify, please go through this article: https://stackoverflow.com/questions/22413009/jasmine-javascript-testing-tobe-vs-toequal

mashrukh-islam replied:
I tried without stringifying, and I used both toBe and toEqual. But I was getting the same error as before.
`Received: serialiazes to the same string`

alichherawalla commented on app/containers/ITunesProvider/reducer.js:
```diff
@@ -0,0 +1,62 @@
+        break;
+      case iTunesProviderTypes.CLEAR_TRACK_DETAILS:
+        return initialState;
+      case iTunesProviderTypes.SUCCESS_GET_TRACK_DETAILS:
```
draft.tracks = [....state.tracks, action.data];
 draft.trackDetails = action.data;

mashrukh-islam replied:
done

mashrukh-islam committed:
additional tests for saga and some refactoring
